app: Add version file and releasing information (#322)
nrfconnect/ncs-serial-modem: doc/app/sm_releasing.rst

.. _sm_releasing:

Releasing |SM|
##############

Before each official release or pre-release, update the version metadata below.
Each published image must carry a higher value than the previous one on the same line (application sign version or MCUboot ``fw_info.version``).

Application — :file:`app/VERSION`
=================================

Sets ``AT#XSMVER`` and the MCUboot application sign version (``CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION``), which is used for application downgrade prevention.
Format: ``major.minor.patchlevel+VERSION_TWEAK``. See `Zephyr application VERSION`_.

Pre-releases
------------

Before General Availability (GA): fix ``VERSION_MAJOR``, ``VERSION_MINOR``, and ``PATCHLEVEL`` on the pre-GA line; increment ``VERSION_TWEAK`` per published build.

> Example for 2.0.0 GA — line ``1.99.0``, images ``1.99.0+1``, ``1.99.0+2``, …

Author: VERSION_TWEAK is not taken into account when doing rollback prevention, so it is possible to update the application from pre-release 2 to pre-release 1.

Reviewer: I don't think this is a problem but worth mentioning in the doc

.. code-block:: none

   VERSION_MAJOR = 1
   VERSION_MINOR = 99
   PATCHLEVEL = 0
   VERSION_TWEAK = 1

General Availability (GA)
-------------------------

Set the product version and ``VERSION_TWEAK = 0``.

.. code-block:: none

   VERSION_MAJOR = 2
   VERSION_MINOR = 0
   PATCHLEVEL = 0
   VERSION_TWEAK = 0

``2.0.0+0`` must be greater than every shipped ``1.99.0+N`` pre-release.

After GA: increment ``PATCHLEVEL`` or ``VERSION_TWEAK`` for each new published application build on that line.

Do not bump ``VERSION_TWEAK`` for unpublished local or CI builds.

> Record Git tag ↔ ``+`` value in release notes.

Author: We should probably release the git tags in this format as well. So that 2.0.0-preliminary would become 1.99.0 (or 1.99.0+1).

.. code-block:: none

   # First v2.0.0 pre-release
   VERSION_MAJOR = 1
   VERSION_MINOR = 99
   PATCHLEVEL = 0
   VERSION_TWEAK = 0

.. code-block:: none

   # Second v2.0.0 pre-release
   VERSION_MAJOR = 1
   VERSION_MINOR = 99
   PATCHLEVEL = 1
   VERSION_TWEAK = 0

.. code-block:: none

   # v2.0.0 release
   VERSION_MAJOR = 2
   VERSION_MINOR = 0
   PATCHLEVEL = 0
   VERSION_TWEAK = 0

``2.0.0+0`` must be greater than every shipped ``1.99.N+0`` pre-release.

Future v2.1.0 GA
~~~~~~~~~~~~~~~~

Pre-releases before v2.1.0 GA, use the ``2.0.99`` line (for example ``2.0.99``, then ``2.0.100``), incrementing ``PATCHLEVEL`` for each published pre-release, following the same pattern as ``1.99.0``, ``1.99.1`` before v2.0.0 GA.
At v2.1.0 GA, set ``VERSION_MAJOR``, ``VERSION_MINOR``, and ``PATCHLEVEL`` to ``2``, ``1``, and ``0``.

General Availability and later
------------------------------

Set the product version and ``VERSION_TWEAK = 0`` at GA.
After GA, increment ``PATCHLEVEL`` for each new published application build on that line.

Do not bump version fields for unpublished local or CI builds.
Record Git tag ↔ :file:`app/VERSION` in release notes.

> Confirm the version with :ref:`SM_AT_XSMVER`.

Reviewer 2: ```suggestion
The following guidelines describe how to manage versioning for General Availability and later:

* Set the product version and ``VERSION_TWEAK = 0`` at GA.
* After GA, increment ``PATCHLEVEL`` for each new published application build on that line.
* Do not update version fields for unpublished local or CI builds.
* Record the mapping between Git tag and :file:`app/VERSION` in the release notes,
* Confirm the version with SM_AT_XSMVER.
```
Maybe we could convert these to bullet points just for more readability.

.. _sm_releasing:

Releasing |SM|
##############

> Before each official release or pre-release, update the version metadata below.

Reviewer 2: ```suggestion
Before each official release or pre-release, update the following version metadata.
```

Each published application image must carry a higher MCUboot sign version than the previous one on the same product line.

> Increment ``CONFIG_FW_INFO_FIRMWARE_VERSION`` only when the MCUboot bootloader image changes.

Reviewer: I'd probably leave some of this text out as it looks more like internal notes when doing the releasing.
But let's not spend time tuning the text too much.

Application — :file:`app/VERSION`

> =================================

Reviewer 2: ```suggestion
Application - :file:`app/VERSION`
=================================
```

:file:`app/VERSION` is the single source of truth for:

* The ``major.minor.patchlevel`` prefix in ``AT#XSMVER`` ``<sm_version>`` (Git-describe style: ``v`` + that triplet + optional ``-N-g<hash>`` from Git).
* The MCUboot application sign version (``CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION``, ``major.minor.patchlevel+VERSION_TWEAK``), used for application downgrade prevention during DFU.

See `Zephyr application VERSION`_.

DFU and ``AT#XSMVER``
---------------------

For application firmware update, a host can read ``AT#XSMVER`` before and after DFU and compare the ``major.minor.patchlevel`` prefix of ``<sm_version>`` (the part from :file:`app/VERSION`, before any ``-N-g<hash>`` Git suffix).

> That value must increase when moving to a newer published image; MCUboot enforces the same ordering via the sign version derived from :file:`app/VERSION`.

Reviewer 2: ```suggestion
That value must increase when moving to a newer published image.
MCUboot enforces the same ordering using the sign version derived from :file:`app/VERSION`.
```

The Git suffix is informational only and is not used for downgrade checks.

Pre-release numbering
---------------------

Before General Availability (GA) for a target ``X.Y.Z``, use a staging line below ``X.Y.Z`` and increment ``PATCHLEVEL`` for each published pre-release.

> Keep ``VERSION_TWEAK = 0`` unless you have a separate reason to use it after GA.

Reviewer 2: ```suggestion
Keep ``VERSION_TWEAK = 0`` unless you have a specific reason to use it after GA.
```

> 2.0 GA

Reviewer: There are more of these in the PR but generally I would like 2.0 to be v2.0.0.

```suggestion
v2.0.0 GA
```

~~~~~~~~~

The first v2.0.0 pre-release uses ``1.99.0``, the second ``1.99.1``, and so on.
At GA, set the product version to ``2.0.0``.

.. code-block:: none

   # First v2.0.0 pre-release
   VERSION_MAJOR = 1
   VERSION_MINOR = 99
   PATCHLEVEL = 0
   VERSION_TWEAK = 0

.. code-block:: none

   # Second v2.0.0 pre-release
   VERSION_MAJOR = 1
   VERSION_MINOR = 99
   PATCHLEVEL = 1
   VERSION_TWEAK = 0

.. code-block:: none

   # v2.0.0 release
   VERSION_MAJOR = 2
   VERSION_MINOR = 0
   PATCHLEVEL = 0
   VERSION_TWEAK = 0

``2.0.0+0`` must be greater than every shipped ``1.99.N+0`` pre-release.

Future v2.1.0 GA
~~~~~~~~~~~~~~~~

> Pre-releases before v2.1.0 GA use the ``2.0.99`` line (for example ``2.0.99``, then ``2.0.100``), incrementing ``PATCHLEVEL`` for each published pre-release—the same pattern as ``1.99.0``, ``1.99.1`` before v2.0.0 GA.

Reviewer 2: ```suggestion
Pre-releases before v2.1.0 GA, use the ``2.0.99`` line (for example ``2.0.99``, then ``2.0.100``), incrementing ``PATCHLEVEL`` for each published pre-release, following the same pattern as ``1.99.0``, ``1.99.1`` before v2.0.0 GA.
```

At v2.1.0 GA, set ``VERSION_MAJOR``, ``VERSION_MINOR``, and ``PATCHLEVEL`` to ``2``, ``1``, and ``0``.

General Availability and later
------------------------------

Set the product version and ``VERSION_TWEAK = 0`` at GA.
After GA, increment ``PATCHLEVEL`` for each new published application build on that line.

Do not bump version fields for unpublished local or CI builds.
Record Git tag ↔ :file:`app/VERSION` in release notes.

> Confirm the version with :ref:`SM_AT_XSMVER`.

Reviewer 2: ```suggestion
Do not update version fields for unpublished local or CI builds.
Record the mapping between Git tags and app/VERSION in the release notes, and confirm the version using ``SM_AT_XSMVER``.
```

Author: The checking of the version with AT#XSMVER and the mapping of the git tags in release notes are two completely different operations. This is kind of a checklist on what to do and what not to do.

* Record the mapping between Git tag and :file:`app/VERSION` in the release notes,

> * Confirm the version with SM_AT_XSMVER.

Reviewer 2: ```suggestion
* Confirm the version with ``SM_AT_XSMVER``.
```

=============================================

:file:`app/sysbuild/mcuboot/prj.conf`

Increment only when the MCUboot bootloader image changes.

> Used by NSIB (B0) for MCUboot slot downgrade prevention — independent of :file:`app/VERSION`.

Reviewer 2: ```suggestion
In :file:`app/sysbuild/mcuboot/prj.conf`, increment the value only when the MCUboot bootloader image changes.
The value is used by NSIB (B0) for MCUboot slot downgrade prevention and is independent of :file:`app/VERSION`.
```
Please check the meaning.

After an MCUboot update, ``AT#XBOOTINFO=0`` reports the active slot value.

> See :ref:`serial_modem_release_notes`.

Reviewer 2: ```suggestion
See :ref:`serial_modem_release_notes` for more information.
```
